Propagate status codes from shard failures appropriately

pawankartik-elastic opened the pull request:
If the underlying error caused is `NodeNotConnectedException` and the reason is `connection already closed`, the status code that should be returned is 502 rather than 500. This allows Kibana to retry the search op.

Edit: the scope of this PR itself has changed since I first worked on it. The goal of this PR is to let the status code propagate so that it is returned back to the user appropriately. There's a separate task to map out `Exception`-s to `502`; hence I'll be removing the related code form this PR. Adding this edit to let the readers understand how the scope has evolved.

Relates ES-10191.
Fixes https://github.com/elastic/elasticsearch/issues/118482.

pawankartik-elastic committed:
Return 502 if the underlying error is `NodeNotConnectedException`

pawankartik-elastic committed:
Traverse through the cause stack trace and check for `NodeNotConnectedException` and undo `status()` override in `NodeDisconnectedException`

pawankartik-elastic committed:
Rewrite `while` condition

pawankartik-elastic committed:
Fix: precommit

DaveCTurner commented on server/src/main/java/org/elasticsearch/ExceptionsHelper.java:
```diff
@@ -91,7 +92,18 @@ public static Throwable unwrapCause(Throwable t) {
         }
-        return result;
+
+        // Traverse through the stacktrace and check if the error is caused by NodeNotConnectedException.
```
We shouldn't be unwrapping things other than `ElasticsearchWrapperException` here. Do we ever see this exception wrapped inside something that isn't an `ElasticsearchWrapperException`? If so, we should fix that wrapper.

pawankartik-elastic replied:
If there's an exception that eventually leads to `NodeNotConnectedException` (or a variant of `ConnectTransportException` and I believe there could be many) and it doesn't implement `status()`, won't it end up returning `500` before the propagation even reaches to `NodeNotConnectedException`?

DaveCTurner replied:
Yes, so let's fix that. But not here. This concern applies to all possible root exceptions, not just `NodeNotConnectedException` or `ConnectTransportException`, so it's not appropriate to have some special-case handling for those exceptions here.

- Root exception has wrong `status()` --> fix that exception implementation.
- Wrapper fails to propagate `status()` of inner exception correctly --> fix the wrapper.

> I believe there could be many

I don't think that's the case.

pawankartik-elastic committed:
Let status codes propagate rather than walking the stacktrace explicitly

Walking the stacktrace explicitly and looking for a specific error (node
connection-related errors in this case) is a workaround rather than a
proper fix. Instead, let the status codes propagate all the way to the
top so that they can be reported as-is.

pawankartik-elastic committed:
Fix: unused import

pawankartik-elastic committed:
Fix null deref

pawankartik-elastic committed:
Merge branch 'main' into pkar/node-not-conn-502

DaveCTurner commented on server/src/main/java/org/elasticsearch/transport/NodeDisconnectedException.java:
```diff
@@ -34,4 +36,10 @@ public NodeDisconnectedException(StreamInput in) throws IOException {
     }
+
+    @Override
+    public RestStatus status() {
+        Throwable cause = getCause();
+        return cause != null ? ExceptionsHelper.status(cause) : RestStatus.BAD_GATEWAY;
+    }
```
We have a separate task open to make _all_ descendants of `ConnectTransportException` map onto a 502. You need take no action on this front, and definitely shouldn't be doing something like this to specific subclasses.

DaveCTurner commented on server/src/main/java/org/elasticsearch/transport/NodeNotConnectedException.java:
```diff
@@ -33,4 +34,9 @@ public NodeNotConnectedException(StreamInput in) throws IOException {
     }
+
+    @Override
+    public RestStatus status() {
+        return RestStatus.BAD_GATEWAY;
```
Likewise here.

DaveCTurner reviewed (commented):
I have no strong opinions about the implementation of `SearchPhaseExecutionException#status()`, this is something that the search-foundations team should review.

We shouldn't be changing the `ConnectTransportException` subclasses here.

pawankartik-elastic committed:
Do not map descendants of `ConnectTransportException` to `502`

pawankartik-elastic committed:
Fix: precommit

pawankartik-elastic committed:
Merge branch 'main' into pkar/node-not-conn-502

pawankartik-elastic committed:
Do not account for 4xx status codes

4xx status codes are not likely to appear along with 5xx status codes.
As a result, we do not need to account for them when looking at shard
failures' status codes.

pawankartik-elastic committed:
Remove unnecessary `switch` case

In reference to the previous commit: this case is no longer needed.

pawankartik-elastic committed:
Rewrite code comment

javanna commented on server/src/main/java/org/elasticsearch/action/search/SearchPhaseExecutionException.java:
```diff
@@ -73,15 +73,28 @@ public RestStatus status() {
+                case RestStatus.INTERNAL_SERVER_ERROR:
+                case RestStatus.NOT_IMPLEMENTED:
+                case RestStatus.HTTP_VERSION_NOT_SUPPORTED:
+                case RestStatus.INSUFFICIENT_STORAGE:
```
can you expand on these special cases? The last two are never returned from ES, and the second one only from some very specific places, I wonder why we need to treat it differently. Could we otherwise have some kind of logic like if status == 500 or status > 504 etc. ? Is that the intention?

pawankartik-elastic replied:
The code that's being replaced tracks for any status code as long as it's >= 500. So I retained them here in the `switch` as well. Since you say that we don't return the last 2, should I get rid of them then?

javanna replied:
I see, so you have a preference for mentioning each specific status code? I think it's a matter of preference, I have a hard time connecting those descriptions with the actual status code number, hence I have a slight preference for reasoning about error ranges perhaps. It may be me.

pawankartik-elastic replied:
Since you mentioned that the last 2 are never really returned and the 2nd one only from specific places, I could remove them. I'm now leaning towards this approach. Alternately, I could retain them and simply add a code comment.

pawankartik-elastic replied:
I can change the code to use actual status codes rather than the `enum`-s. In fact, I'm alright with that. And regarding the status codes that aren't used/returned (the ones you've talked in the first comment), should I get rid of them?

javanna commented on server/src/test/java/org/elasticsearch/rest/RestResponseTests.java:
```diff
@@ -505,6 +506,29 @@ public void testSupressedLogging() throws IOException {
+
+        RestResponse response = new RestResponse(channel, new RemoteTransportException("unused message", ex));
+        assertThat(response.status(), is(RestStatus.BAD_GATEWAY));
+    }
```
could you expand to testing to cover more cases? I think we could move these to a SearchPhaseExecutionException specific test as well?

javanna reviewed (commented):
left two comments and questions, thanks!

pawankartik-elastic committed:
Address review comments

Commit:
[CI] Auto commit changes from spotless

pawankartik-elastic committed:
Merge branch 'main' into pkar/node-not-conn-502

javanna reviewed (approved):
LGTM thanks @pawankartik-elastic !

pawankartik-elastic committed:
Update docs/changelog/118016.yaml

elasticsearchmachine commented:
Hi @pawankartik-elastic, I've created a changelog YAML for you.

pawankartik-elastic committed:
Merge branch 'main' into pkar/node-not-conn-502

elasticsearchmachine commented:
Pinging @elastic/es-search-foundations (Team:Search Foundations)

elasticsearchmachine commented:
## 💚 Backport successful
| Status | Branch | Result |
|:------:|:------:|:------:|
| ✅ |  [8.x](https://github.com/elastic/elasticsearch/pull/119205)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/119205">](https://github.com/elastic/elasticsearch/pull/119205) |
